docs/recipes/solid.md

```diff
@@ -0,0 +1,186 @@
+# Usage with SolidJS
+
+XState can be used with SolidJS to:
+
+- Coordinate local state
+- Manage global state
+- Consume data from other hooks
+- etc.
+
+## Local state
+
+```js
+import { useMachine } from '@xstate/solid';
+import { toggleMachine } from '../path/to/toggleMachine';
+
+function Toggle() {
+  const [current, send] = useMachine(toggleMachine);
+
+  return (
+    <button onClick={() => send('TOGGLE')}>
+      {current.matches('inactive') ? 'Off' : 'On'}
+    </button>
+  );
+}
+```
+
+## Global State in Context
+
+### Context Provider
+You can store a machine or service in [SolidJS context](https://www.solidjs.com/docs/latest/api#createcontext) to make them available throughout the component tree. Because @xstate/solid produces machines, actors, and services with SolidJS' fine-grained reactivity, you can access machine context and methods (e.g. `matches` or `can`) without worrying about wasted re-renders
+```js
+import { createContext } from 'solid-js';
+import { useMachine } from '@xstate/solid';
+import { authMachine } from './auth.machine';
+
+export const GlobalStateContext = createContext({});
+
+export const GlobalStateProvider = (props) => {
+  const authService = useMachine(authMachine);
+
+  return (
+    <GlobalStateContext.Provider value={{ authService }}>
```

Comment:
This feels a little bit weird to me - but I'm talking mainly from a perspective of a heavy React user. It might not be weird in the Solid land.

OTOH I wonder if with Solid there is even a need for `useMachine` in its current shape. It feels that in Solid we could just return a `service` that would have reactive `service.state`. So accessing `service.state.context.foo` would subscribe to it granularly etc. What do you think about something like that? The name of the hook that would return this is up for a debate but that is a secondary concern here.

Reply:
It feels a bit weird for me too having mainly used React for 7 years before Solid. This is similar to the React recipe but in Solid we can skip the `useInterpret` step and directly use the machine. There is an example in Solid's docs that use a similar `[state, action]` destructuring pattern within context. [https://www.solidjs.com/docs/latest/api#createcontext](https://www.solidjs.com/docs/latest/api#createcontext)

Using a machine in context could be simplified to this 
```tsx
const {authService: [state, send]} = useContext(GlobalStateContext);
```

Alternatively, you could still use a similar pattern to React with `createService`  in context then use the service with `useActor`.  
```tsx
const { authService } = useContext(GlobalStateContext);
const [state, send] = useActor(authService)
```

The `useMachine` hook could be changed to return a service object. That could come down to developer preference between a hook returning an array with a clear `[state, send, service]` or a service object with `{state, send, ...other Interpreter methods}`.

Reply:
I just wonder if at the end of the day passing around a "solidified" service wouldn't be easier. This way you would only have to pass around a single object and that would already have `state` and `send` available on it. In addition to that, both would naturally already be "scoped" with this `GlobalStateContext` pattern.

The `createService`+`useActor` pattern is also tempting. This way it would look more consistent to other integration packages. But OTOH each one has its own right to have a different API if the host framework provides unique traits that allow for that different API to be better than the "consistent API". After all, almost nobody should use both `@xstate/react` and `@xstate/solid` in the same project - so the difference in the API wouldn't really affect them all that much.

It's really hard to decide for me on any of those APIs because all of them are OK. I will give this a thought over the weekend and discuss it with David after that. We could also ask some XState+Solid devs for opinions here.

Reply:
I played around with the idea today and have a working version that returns a service from `useMachine` with reactive state. It supports both maintaining a "solidified" service as well as destructuring. 

```tsx
// maintaining service structure example, authService has all properties and methods of Interpreter

const authService = useMachine(authMachine);

authService.send('login');

authService.subscribe(state => /* ...do something with state */);

return authService.state.context.username;

// with destructuring

const {state, send, subscribe} = useMachine(authMachine);

send('login');

subscribe(state => /* ...do something with state */);

return state.context.username;
```

This was a quick and somewhat messy implementation just to see how it worked. I didn't tackle renaming any of the hooks and there are several things I can optimize and trim out. I put the changes in another branch, [here's the diff](https://github.com/GoldingAustin/xstate/pull/2/files#diff-964efe5e11f928b6ebe572935dfe4203f1c15a327f3433654c0bd7c3bc3cd490) and a [link to the code](https://github.com/GoldingAustin/xstate/tree/feature/xstate-solidjs-alternate/packages/xstate-solid). All the tests are passing and the API is cleaner (in my opinion).